components/datetime/src/fields/symbols.rs

```diff
@@ -38,6 +38,96 @@ pub enum FieldSymbol {
     TimeZone(TimeZone),
 }
 
+impl FieldSymbol {
+    /// Symbols are necessary components of `Pattern` struct, which
+    /// uses efficient byte serialization and deserialization via `zerovec`.
+    ///
+    /// The `FieldSymbol` impl provides non-public methods that can be used to efficiently
+    /// convert between `u8` and the symbol variant.
+    ///
+    /// The serialization model packages the variant in one byte.
+    ///
+    /// 1) The top four bits are used to determine the type of the field.
+    ///    (Examples: `Year`, `Month`, `Hour`)
+    ///
+    /// 2) The bottom four bits are used to determine the symbol of the type.
+    ///    (Examples: `Year::Calendar`, `Hour::H11`)
+    ///
+    /// # Diagram
+    /// <pre>
+    /// ┌─┬─┬─┬─┬─┬─┬─┬─┐
+    /// ├─┴─┴─┴─┼─┴─┴─┴─┤
+    /// │ Type  │Symbol │
+    /// └───────┴───────┘
+    /// </pre>
+    ///
+    /// # Optimization
+    ///
+    /// This model is optimized to package data efficiently when `FieldSymbol`
+    /// is used as a variant of `PatternItem`. See the documentation of `PatternItemULE`
+    /// for details on how it is composed.
+    ///
+    /// # Constraints
+    ///
+    /// This model limits the available number of possible types and symbols to 16 each.
+
+    #[inline]
+    pub(crate) fn idx_in_range(kv: &u8) -> bool {
+        let symbol = kv & 0b0000_1111;
+        let field_type = kv >> 4;
+        match field_type {
+            0 => Year::idx_in_range(&symbol),
+            1 => Month::idx_in_range(&symbol),
+            2 => Day::idx_in_range(&symbol),
+            3 => Weekday::idx_in_range(&symbol),
+            4 => DayPeriod::idx_in_range(&symbol),
+            5 => Hour::idx_in_range(&symbol),
+            6 => true,
```

Comment:
question: should this be `=> symbol == 0`? ULE types have the (non safety) equality invariant that byte equality is equality, and we lose that if we allow multiple values for the low nibble on `minute`